src/python/pants/engine/README.md

```diff
@@ -76,52 +76,80 @@ In cases where this search detects any ambiguity (generally because there are tw
+1. a compact and high-performance representation which can be stably shared across FFI boundaries,
+2. static type checking via [the `dataclasses` mypy plugin](https://github.com/python/mypy/blob/master/mypy/plugins/dataclasses.py),
+3. a concise, standard, and Pythonic way to declare classes.
+
+
+#### Example Usage of `@dataclass` for Engine `Param`s
+
+*Note that the [3rdparty `dataclasses` library](https://github.com/ericvsmith/dataclasses) must be in your BUILD file and `import`ed if you're running Pants with Python 3.6!*
 
 ```python
-class FormattedInt(datatype(['content'])): pass
+from dataclasses import dataclass
+
+from pants.util.meta import frozen_after_init
+
+# Pants requires that engine Params have a stable hash. This can be accomplished with the
+# `frozen=True` argument set in the `@dataclass` decorator.
+@dataclass(frozen=True)
+class FormattedInt:
+  content: str
 
 @rule
 def int_to_str(value: int) -> FormattedInt:
-  return FormattedInt('{}'.format(value))
+  return FormattedInt(str(value))
 
-# Field values can be specified with positional and/or keyword arguments in the constructor:
-x = FormattedInt('a string')
-x = FormattedInt(content='a string')
+# `@dataclass` objects can be easily inspected:
+print(x)                        # 'FormattedInt(content="a string")'
 
-# Field values can be accessed after construction by name or index:
-print(x.content)    # 'a string'
-print(x[0])         # 'a string'
+# All MyPy types, including parameterized types, may be used in field definitions. The runtime type
+# must still be hashable.
+@dataclass(frozen=True)
+class TypedDatatype:
+  """Example of a datatype with a more complex field type constraint."""
+  field_name: Union[str, int]
 
-# datatype objects can be easily inspected:
-print(x)            # 'FormattedInt(content=a string)'
+print(TypedDatatype("huh")) # 'TypedDatatype(field_name=huh)'
 ```
 
-#### Types of Fields
+#### Using `@frozen_after_init` with `@dataclass`es
 
-`datatype()` accepts a list of *field declarations*, and returns a type which can be subclassed. A
-*field declaration* can just be a string (e.g. `'field_name'`), which is then used as the field
-name, as with `FormattedInt` above. A field can also be declared with a tuple of two elements: the
-field name string, and a `TypeConstraint` for the field (e.g. `('field_name',
-Exactly(FieldType))`). The bare type name (e.g. `FieldType`) can also be used as a shorthand for
-`Exactly(FieldType)`. If the tuple form is used, the constructor will create your object, then raise
-an error if the field value does not satisfy the type constraint.
+`@frozen_after_init` can also be used with `@dataclass(unsafe_hash=True)` to create engine `Param`s which can modify their fields within the `__init__()` method. This is useful if you want to perform any normalization on the args to `__init__()`, such as allowing `List` and then casting this to an immutable `Tuple` in `__init__()`:
```

Comment:
Great! I recommend inverting the order of these sentences so that you lead with the _why_ and then follow up with the _how_.

> Often, you'll want a custom `__init__()` but to still have the dataclass be marked as frozen; for example, you may want to allow the user to pass `Iterable[str]` in the constructor and then to convert it into an immutable `Tuple[str, ...]`.
>
> Dataclasses do not normally allow you to mix custom `__init__()` implementations with `frozen=True`, so we created a utility called `@frozen_after_init` to get the best of both worlds.